### PR TITLE
Fix problem with not proper calculated loss function based

### PR DESCRIPTION
**Pull request by machineko**
on wrong len regulating function.

### CODE REVIEW

**Comment by machineko** [comment]
@dathudeptrai

**Comment by dathudeptrai** [comment]
@machineko I do not think this is a bug. Let me explained why there is a length mismatch problem between ground-truth duration and predicted duration.

Assume we have `2` GPU and the `batch_size` for each GPU is 32 then the `global_batch_size` is 64. In 64 those samples, the `maximum length` of mel is 800 then the `maximum sum duration` is also 800. After that, we apply `experimental_distributed_dataset` for the data loader and the `maximum length` of mel is still 800 but the `maximum sum duration`  will be `<=` 800 in each GPU. To resolve this mismatch we need to slice the `ground truth duration` to make its length equal to the predicted duration length as the code in the repo did.

@machineko

**Comment by machineko** [comment]
Predicted duration can be shorter then GT from nature of len regulator in for example fastspeech it happen especialy in early epoches u can try it on single gpu (thats how i was testing) and see on for example Vctk i got like 10/20% of predictions that are shorter then GT

**Comment by dathudeptrai** [comment]
> Predicted duration can be shorter then GT from nature of len regulator in for example fastspeech it happen especialy in early epoches u can try it on single gpu (thats how i was testing) and see on for example Vctk i got like 10/20% of predictions that are shorter then GT

i do not think so, the length of predicted duration `MUST` be equal to the len of groundtruth duration. The value that shorter is the `sum_duration` not the length.

**Comment by machineko** [comment]
Ill debug it again with some results but im prety sure there was a lot of times when missmatch was caused by smaller predictions then GT and will be back with it then 😄 (Else i wouldnt fix it xd)

**Comment by dathudeptrai** [comment]
> Ill debug it again with some results but im prety sure there was a lot of times when missmatch was caused by smaller predictions then GT and will be back with it then  (Else i wouldnt fix it xd)

thanks, :D, pls give me a sample that causes the mismatch bug :D. I never get this bug in a single GPU.

**Comment by machineko** [comment]
len_missmatch

As show above length will be different cause of len regulation function so GT MEL and Predicted MEL will be diffrent (This is on VCTK dataset).

And I found 2 reasons for that.

First and most common => len generated from MFA/Taco2 != len from GT and mismatch can be substantial in some cases so removing ~20/30 frames from loss function are causing big problems with bad endings without <end> token.

Second => len generated from learned  duration_predictor is different  (so you should punish it not encourage :P [I'm using duration prediction in training after my dur_pred loss stagnate] )

**Comment by dathudeptrai** [comment]
@machineko 

for a first reason, I think user need to check by themself to make sure that the generated duration from Tacotron2/MFA match exactly the length of the ground truth mel-spectrogram. 

fro a second reason, after ur duration loss staturated, you can use the output of duration predictor for training with `tf.stop_gradient`. 

```
...
duration_predicted = ...
duration_predicted_round = tf.stop_gradient(tf.round(duration_predicted))
duration_predicted_round = posprocess_fn(...)  ---> to make sure tf.reduce_sum(duration_predicted_round) == len(mel_gts)
.... # Use the duration_predicted_round above as duration_gts for training.
```

**Comment by machineko** [comment]
But why should i if i can just fix loss function to work properly and prepare it to work properly with XLA?
Also u cant make sure sum of lens from MFA/Taco == mel len without padding in prepossessing or trimming (which is done right now) end of audio and dropping a lot of examples with big differences in sum(lens) != mel_len , its a lot more complicated then just fixing loss function and let model train on proper mels and unfreeze duration training to learn true duration's values not forced by MFA/Taco2.

As my main models are written in Flax and its not FS2 but still use len_regulator i found this to be one of the only working solutions for noisy datasets as MFA kinda sucks with it :)
This is just small contribution for easier and better quality models.
But I wont be forcing any solution.

**Comment by dathudeptrai** [comment]
> But why should i if i can just fix loss function to work properly and prepare it to work properly with XLA?
> Also u cant make sure sum of lens from MFA/Taco == mel len without padding in prepossessing or trimming (which is done right now) end of audio and dropping a lot of examples with big differences in sum(lens) != mel_len , its a lot more complicated then just fixing loss function and let model train on proper mels and unfreeze duration training to learn true duration's values not forced by MFA/Taco2.
> 
> As my main models are written in Flax and its not FS2 but still use len_regulator i found this to be one of the only working solutions for noisy datasets as MFA kinda sucks with it :)
> This is just small contribution for easier and better quality models.
> But I wont be forcing any solution.

let me think :)) and fix the loss function since we have many situations here :)).
